data_transfer/devices/byteflies.py

```diff
@@ -0,0 +1,227 @@
+        ...
+
+        NOTE/TODO: will run as BATCH job.
+        """
+        # Note: includes metadata for ALL data records, therefore we must filter them
+        all_records = byteflies_api.get_list(
+            self.session, config.byteflies_group_ids[study_site], from_date, to_date
+        )
+        # utils.write_json("./byteflies.json", all_records)
+        # all_records = utils.read_json(Path("./byteflies.json"))
+
+        # Only add records that are not known in the DB based on stored filename
+        unknown_records = [
+            r for r in all_records if r["IDEAFAST"]["hash"] not in set(all_hashes())
+        ]
+
+        for item in unknown_records:
+            # Pulls out the most relevant metadata for this recording
+            recording = self.recording_metadata(item)
+
+            # NOTE: lookup in .csv export from inventory TODO: translate to inventory api
+            if not (
+                resolved_device_id := byteflies_api.serial_by_device(recording.dot_id)
+            ):
+                print(f"Unknown Device:\n   {recording}")
+                continue  # Err: SKIP RECORD
+            else:
+                resolved_device_id = utils.format_id_device(resolved_device_id)
+
+            # saving resolved patient_id back in recording for logging
+            recording.patient_id = (
+                recording.patient_id  # very sporadically the record has a patient_id already
+                or self.__patient_id_from_ucam(
+                    resolved_device_id, recording.start, recording.end
+                )
+                or self.__patient_id_from_inventory(
+                    resolved_device_id, recording.start, recording.end
+                )
+            )
+
+            if not (
+                resolved_patient_id := utils.format_id_patient(recording.patient_id)
+            ):
+                print(f"Unknown Patient:\n   {recording}")
+                continue  # Err: SKIP RECORD
+
+            record = Record(
+                # can relate to a single download file or a group of files
+                hash=item["IDEAFAST"]["hash"],
+                manufacturer_ref=recording.recording_id,
+                device_type=self.device_type,
+                device_id=resolved_device_id,
+                patient_id=resolved_patient_id,
+                start_wear=recording.start,
+                end_wear=recording.end,
+                meta=dict(
+                    studysite_id=recording.group_id,
+                    recording_id=recording.recording_id,
+                    signal_id=recording.signal_id,
+                    algorithm_id=recording.algorithm_id,
+                ),
+            )
+
+            create_record(record)
+            print(f"Record Created:\n   {record}")
+
+            del item["IDEAFAST"]  # remove record-specific temporary metadata
+            path = Path(config.storage_vol / f"{record.manufacturer_ref}-meta.json")
```

Comment:
Note: this stores the "meta.json" inside `/input/` rather than `/input/{resolved_patient_id}/`. At this point then, it would be worth making the folder `/{resolved_patient_id}/{resolved_device_id}/`

Perhaps a helper method (in `utils.py`) as this will be used in `download_file` and the same for all pipelines?

Reply:
Good suggestion. I've added this as a helper method for the `Record` class:

```python
    def download_folder() -> str:
        """consistenly structure target folder for download"""
        return f"{device_type}/{self.patient_id}/{self.device_id}"
```